### ISSUE TITLE
Add Delimiter or NewLine argument to Set-Content

### ISSUE THREAD

**Issue by me-kell** [PowerShell/PowerShell#23906]
### Summary of the new feature / enhancement

`Set-Content` adds per default a Windows Newline, i.e. `CR LF`.

With `-NoNewline` it writes without inserting spaces or new lines between the output strings.

There are use cases that the output file should have other line endings. E.g. `LF` for Unix or `CR` for Mac.

Why not have an argument to pass the type of newline (`CR LF`, `LF`, `CR` or whatever needed)?

On the other side `Get-Content` has the parameter `-Delimiter`. Why not make `Set-Content` a converse cmdlet to `Get-Content` and adding the parameter `-Delimiter` to `Get-Content`?

**Comment by rhubarb-geek-nz** [comment]
> `Set-Content` adds per default a Windows Newline, i.e. `CR LF`.

My tests show it uses the platform specific line ending [Environment]::NewLine

```
PS>  'foo','bar' | Set-Content -LiteralPath foo
PS> Format-Hex -path foo

   Label: /home/rogerb/tmp/foo

          Offset Bytes                                           Ascii
                 00 01 02 03 04 05 06 07 08 09 0A 0B 0C 0D 0E 0F
          ------ ----------------------------------------------- -----
0000000000000000 66 6F 6F 0A 62 61 72 0A                         foo�bar�
```

I suggest that Set-Content's job should be to simply write a file. rather than having multiple options for formatting, those can be done in another pipeline stage.

**Comment by me-kell** [comment]
@rhubarb-geek-nz I agree with you that `Set-Content` and `Out-File` should not have multiple options for formatting.

But writing to files for the platform where the script runs is not the only use case.

Mostly we develop files for other platforms (Linux, Windows etc.). Those files are in different repositories and have their own line endings.

Use case: managing configuration files for different virtual machines. It is painful to have to convert the line endings for every file before we commit the files or copy the files to the target system.

And it makes no sense to redefine `[Environment]::NewLine` for every file output.

It is hard to understand why not to add a `-NewLine` parameter that will be passed to callback [OnWriteLine of Out-String](https://github.com/PowerShell/PowerShell/blob/v7.4.2/src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/out-string/Out-String.cs#L117-L134). Simply passing the `newLineDelimiter` to `_buffer.Append(s, newLineDelimiter);` if defined or otherwise using `_buffer.AppendLine(s);`

```cs
        private void OnWriteLine(string s)
        {
            if (_stream)
            {
                this.WriteObject(s);
            }
            else
            {
                if (_noNewLine)
                {
                    _buffer.Append(s);
                }
                else
                {
                    _buffer.AppendLine(s);
                }
            }
        }
```

**Comment by mklement0** [comment]
@me-kell, I second your suggestion, which in fact was first proposed 7 years ago:

* #3855

The "won't fix" decision accompanying the closure was based on arguing that the `-join` operator could be used - a rationale that understandably proved to be quite unpopular.

I'm hoping the issue can be revisited now. /cc @SteveL-MSFT 

---

@rhubarb-geek-nz, in a nutshell: (With the FileSystem provider) `Set-Content` serializes non-string objects  by their `.ToString()` values, whereas `Out-String` uses the formatting system, i.e. serializes objects using the same rich for-display formatting you see in the terminal.

**Comment by rhubarb-geek-nz** [comment]
Yes it would be easy to have the option added, but you still have to work today.

> before we commit the files

The native GIT format is LF. I recommend that you check-in all files to GIT so they are stored with LF. Fortunately this can easily be done by configuring autocrlf. Then the files have the correct line endings for the platform you are on. Similar with SVN, use svn:eol-style native. This lets the source control manage text files far better.

> copy the files to the target system.

Traditionally, eg going back to the days of FTP, doing the CRLF conversions was part of the PUT or GET, so that the file was correct for the recipient. 

Likewise unzip can convert the files to be correct on the system it is unpacked on.

I find it easier to keep files native and standard using the host character set, locale, file ending etc and then it is the packaging or the transfer that knows the destination that converts the text file formats, or puts them in a neutral form for unpacking.

**Comment by me-kell** [comment]
Why is it so difficult to understand the need to parametrize the line endings when writing to a file?

AFAICS there is everywhere plenty of complaints since years about PowerShell missing this point.

Most of script and programming languages offer that. Why must PowerShell be so reluctant to implement this? Is PowerShell really interested in listen to the need of the community?

**Comment by mklement0** [comment]
I hear you, @me-kell - I've been wanting this for 7 years.

While everyone is free to comment on issues, not everyone's comments have equal weight (which is not a reflection on the merit of any given comment), and neither @rhubarb-geek-nz nor I speak in any official capacity, nor do we have any say in what does and doesn't get fixed / implemented.

It is the WGs (working groups) that decide the fate of issues, and we'll have to wait for the relevant WG's feedback.

Unfortunately, though, not all issues are ultimately adjudicated by WGs, as that is contingent on the issue having been labeled for assignment to a specific WG, e.g. `WG-Cmdlets` - which should happen as a result of the triage eventually performed based on the `Needs-Triage` label.

Unfortunately, however, even a WG-specific label may not be enough, as there's now an additional label, `WG-NeedsReview`, suggesting that a WG-specific label _alone_ may not enough to get an issue adjudicated.

Unless I'm missing something, this appears to be a regrettably non-transparent gatekeeping mechanism

**Comment by rhubarb-geek-nz** [comment]
> Most of script and programming languages offer that.

PowerShell is built on .NET tends to follow the capabilities it offers directly. As per the code fragment you posted

[StringWriter.AppendLine](https://learn.microsoft.com/en-us/dotnet/api/system.io.stringwriter.writeline?view=net-8.0) and [StringBuilder.AppendLIne](https://learn.microsoft.com/en-us/dotnet/api/system.text.stringbuilder.appendline?view=net-8.0) both use the default line terminator.

> AFAICS there is everywhere plenty of complaints since years about PowerShell missing this point.

This is where [PSGallery](https://www.powershellgallery.com/) fits in. PowerShell itself does not have to solve every problem. If there was a demand, then there would be something there such as [EndOfLine](https://www.powershellgallery.com/packages/EndOfLine/1.3.1).  

It really is not a show stopper to write your own tool that handles this if it does not already exist.

**Comment by rhubarb-geek-nz** [comment]
[Join-String](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/join-string?view=powershell-7.4) already does what you want.

```
PS> 'foo','bar' | Join-String -OutputPrefix A -Separator B  -OutputSuffix C
AfooBbarC
```

This can be put in a pipeline before Set-Content with -NoNewline.

There is no need to duplicate the behaviour of Join-String wthin Set-Content

**Comment by mklement0** [comment]
@me-kell, continuing the meta discussion:

* Situationally, you may choose _not_ to engage with (non-official) feedback, if you perceive it to be a distraction and suspect that responding to it may only amplify the distraction.

* In other cases, _specious_ feedback may warrant a response; see below for an example.

---

> PowerShell is built on .NET tends to follow the capabilities it offers directly.

PowerShell is - first and foremost - a _shell_, and its design goals differ from the general-purpose .NET framework that underlies it.
While haphazard deviations from .NET APIs are undoubtedly to be avoided, there are fundamental by-design differences (e.g., PowerShell is by default, case-_insensitive_ and mostly culture-_invariant_, unlike .NET). 

As such, in the context of PowerShell, any appeal to what .NET does _by default_ is a specious argument.

>  If there was a demand

The linked issue, #3855, and its predecessor, #2145, have a combined 39 up-votes / hearts as of this writing. By the standards of this repo, this _does_ signal demand.

>  This is where [PSGallery](https://www.powershellgallery.com/) fits in.
> [Join-String](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/join-string?view=powershell-7.4) already does what you want.

Suggesting that something can be done _with custom code_ / by combining _multiple commands_  without acknowledging the convenience / symmetry / performance the issue is asking for is a specious argument.

Case in point:

If `Join-String` were to be used, you'd need:

```powershell
$newline  = "`n"
Set-Content -NoNewLine out.txt -Value (
  'foo', 'bar' | Join-String -Separator $newline -OutputSuffix $newline
) 
```

Needless to say, this approach is obscure and cumbersome, and prevents _streaming_ input.

Contrast it with what this issue asks for, which is (a) convenient, (b) performant, and (c) consistent with `Get-Content`:

```powershell
# Wishful thinking.
'foo', 'bar' | Set-Content out.txt -Delimiter "`n"
```

**Comment by rhubarb-geek-nz** [comment]
> Suggesting that something can be done with custom code 

Heave forfend that somebody uses custom code to solve a problem.

> Needless to say, this approach is obscure, cumbersome, and hurts performance.

A suggested workaround can't hurt performance if the proposed feature does not actually exist. It does have the advantage of can be used right now. It can also be put in its own cmdlet/function Set-ContentWithDelimiter, and use that instead of Set-Content, then when you get the new feature you can alias Set-ContentWithDelimiter to Set-Content. You could use a ForEach-Object within the function to permit streaming.

**Comment by jhoneill** [comment]
> I suggest that Set-Content's job should be to simply write a file. rather than having multiple options for formatting, those can be done in another pipeline stage.

```
#508 [~~€] PS7 ~\Documents\PowerShell> $d = get-item .
#509 [~~€] PS7 ~\Documents\PowerShell> $d

    Directory: C:\Users\james\Documents

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
d----          09/06/2024    15:57                PowerShell

#510 [~~€] PS7 ~\Documents\PowerShell> $d.ToString()
C:\Users\james\Documents\PowerShell
#511 [~~€] PS7 ~\Documents\PowerShell> set-Content myfile -Value $d
#512 [~~€] PS7 ~\Documents\PowerShell> type myfile
C:\Users\james\Documents\PowerShell
#513 [~~€] PS7 ~\Documents\PowerShell> $d | out-file myfile2
#514 [~~€] PS7 ~\Documents\PowerShell> type myfile2

    Directory: C:\Users\james\Documents

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
d----          09/06/2024    15:57                PowerShell
```

Where `Out-File` or the `>`/`>>` operators run an object through the formatter `Set-content` just does `.tostring()` on objects,
It then puts `[environment]::newline` after each string, this CR-LF on Windows and LF on linux. 
There isn't a convenient way to say "On windows output linux line breaks" or vice versa. It has been asked for before.

**Comment by me-kell** [comment]
BTW: `Set-Content` doesn't handle `-Delimiter` in a consistent way. See https://github.com/PowerShell/PowerShell/issues/23907.

> [Join-String](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/join-string?view=powershell-7.4) already does what you want.
> 
> ```
> PS> 'foo','bar' | Join-String -OutputPrefix A -Separator B  -OutputSuffix C
> AfooBbarC
> ```
> 
> This can be put in a pipeline before Set-Content with -NoNewline.
> 
> There is no need to duplicate the behaviour of Join-String wthin Set-Content

But `Join-String` converts the array to a string. This is not optimal with large files.

One could of course do the following:

```
Get-Content input.txt | % { $_ + "`n"} | Set-Content output.txt -NoNewline
```

But, again, why not implement it in `Set-Content` or `Write-File`?

**Comment by me-kell** [comment]
> [Join-String](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/join-string?view=powershell-7.4) already does what you want.
> 
> ```
> PS> 'foo','bar' | Join-String -OutputPrefix A -Separator B  -OutputSuffix C
> AfooBbarC
> ```
> 
> This can be put in a pipeline before Set-Content with -NoNewline.
> 
> There is no need to duplicate the behaviour of Join-String wthin Set-Content

But `Join-String` converts the array to a string. This is not optimal with large files.

One could of course do the following:

```
Get-Content input.txt | % { $_ + "`n"} | Set-Content output.txt -NoNewline
```

But, again, why not implement it in `Set-Content` or `Write-File`?

**Comment by me-kell** [comment]
See the problematic results of hard coding the delimiter and not honouring this parameter https://github.com/PowerShell/PowerShell/issues/23907#issuecomment-2166040816.

**Comment by SteveL-MSFT** [comment]
@PowerShell/wg-powershell-cmdlets reviewed this, we agreed to add to `Set-Content` where it would have symmetry and utility with `Get-Content`, however, we think it would be unnecessary and potentially confusing to users to also add it to `Out-File` and `Out-String` where there are other ways to handle the same problem, but expect most users to use `Set-Content` for this scenario.

**Comment by ArmaanMcleod** [comment]
Would `-Delimiter` always need to be passed with `-NoNewLine`? Or would we allow both or either to be specified like:

- If just `-Delimiter`, concatenate strings using delimiter.
- If just `-NoNewline`, stick with existing behaviour and concatenate strings with no newlines.
- If `-Delimiter` and `-NoNewline` used together, honour the delimiter only and concatenate the strings. E.g. `"a", "b" | Set-Content -Delimiter "," -NoNewLine` would give `a,b`.  Ignore the NoNewLine parameter since Delimiter overrides it.

I guess it just looks weird if someone can do:

```pwsh
"a", "b" | Set-Content -Delimiter "`n" -NoNewLine
```

So maybe easier to make sure they don't get used together.

**Comment by mklement0** [comment]
> may be easier to make sure they don't get used together.

Indeed, it makes no sense to combine `-Delimiter` with `-NoNewLine`, because the latter requests that _no_ "delimiters" (terminators) be used _at all_, so they should be in different parameter sets.

However, an interesting question is whether the `-Delimiter` argument should be treated as a _separator_ (only _between_ elements) or a _terminator_ (to be placed _after each element_, and therefore also the _last_ one), as argued in  #3855:

* Only with the latter would you be able to create, say, Unix-format LF-only text files with ``-Delimiter "`n"`` that exhibit the expected trailing newline behavior.

* To also support _separator_ behavior, the proposed `-NoTrailingNewLine` switch (#5108) could be given a `-NoTrailingDelimiter` _alias_ and omit a trailing "delimiter", whether the default or a given one.

**Comment by ArmaanMcleod** [comment]
@mklement0 Makes sense. 

I think to support both use cases it makes sense to treat `-Delimiter` as a terminator and place after each element, which solves the problem reported by the OP directly.

It also makes sense to also include a `-NoTrailingNewLine`  with a `-NoTrailingDelimiter` alias to also get the separator behaviour. There are use cases where someone would want to keep separator behaviour like `a,b,c` instead of being stuck with `a,b,c,`. I wonder if its just simpler to keep these parameters separate for clarity?

**Comment by mklement0** [comment]
> keep these parameters separate for clarity?

Makes sense - we'll need a new parameter set anyway, so may as well create a dedicated switch.

**Comment by jimmy-zhening-luo** [comment]
I'm stoked on this tbh <3 @ArmaanMcleod
